lib/Sema/PreCheckExpr.cpp

```diff
@@ -744,6 +744,20 @@ Expr *TypeChecker::resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE,
       BaseExpr = TypeExpr::createImplicitForDecl(
           UDRE->getNameLoc(), NTD, BaseDC,
           DC->mapTypeIntoContext(NTD->getInterfaceType()));
+    } else if (Base->getBaseIdentifier().is("self") && Base->getDeclContext()->getContextKind() == DeclContextKind::AbstractClosureExpr) {
+      // If this is an implicit self call, we don't actually
+      // know if self is Self or Optional<Self> (e.g. in a weak self
+      // closure before self is unwrapped).
+      //  - Since we don't know the correct type of self yet,
+      //    we leave this as an UnresolvedDeclRefExpr that is
+      //    populated with the actual type later.
+      //  - We only do this within a closure context, since that's
+      //    the only place where self can be rebound.
+      auto selfNameRef = new DeclNameRef(Base->getBaseIdentifier());
+      BaseExpr = new (Context) UnresolvedDeclRefExpr(*selfNameRef,
+                                                     DeclRefKind::Ordinary,
+                                                     UDRE->getNameLoc());
+      BaseExpr->setImplicit();
```

Comment:
Do you need this UDRE anchor? Seems like the pattern match in CSDiagnostics could drop the outermost conditionals without it.  Generally, passing data across the constraint system via AST construction is a shaky pattern.

Reply:
This change fixes type inference for inferred `self` of `weak self` captures before `self` is unwrapped.

For example, in this case:

```swift
{ [weak self] in
  // implicit self should not be allowed, since self has not been unwrapped
  dismiss()
}
```

The default `DeclRefExpr` created here always has a non-optional type, which made it so the example above would actually compile. This change removes the default non-optional type of the default implicit `DeclRefExpr` by replacing it with a new `UnresolvedDeclRefExpr`, that doesn't have type information yet. The correct type (in this case, an optional) gets populated in a later pass.

I posted about this on the Compiler Development forums, for some more context / background: https://forums.swift.org/t/approaches-for-rejecting-implicit-self-when-self-is-optional/54175. Do you have another approach in mind that would be better / cleaner?